app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/upload/FileUpload.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.cdsreimbursementclaimfrontend.controllers.upload
+
+import cats.implicits.{catsSyntaxEq, catsSyntaxOptionId}
+import com.google.inject.{Inject, Singleton}
+import play.api.Configuration
+import play.api.mvc.Call
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.answers.SupportingEvidencesAnswer
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.upscan.{ScheduledDocument, SupportingEvidence, UploadReference, UpscanUpload}
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.controllers.supportingevidence.{routes => evidenceRoutes}
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.JourneyStatus.FillingOutClaim
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.upscan.UpscanCallBack.UpscanSuccess
+
+trait FileUpload[A] {
+
+  def reviewPage: Call
+
+  def uploadErrorPage: Call
+
+  def handleUploadCallback(uploadReference: UploadReference): Call
+
+  def scanSuccessPage(uploadReference: UploadReference): Call
+
+  def scanErrorPage: Call
+
+  def hasReachedUploadThreshold(maybeAnswer: Option[A]): Boolean
```

Comment:
how does this pick up the different configuration for the different controller classes? for example, users should only be able to upload a maximum of one document but for the supporting evidence the user should be able to upload a maximum of three (I think from memory)?

Reply:
Within the same file `FileUpload.scala`, the implicit file upload instances for concrete answer are defined in the `FileUploadServices` class. 
Which is something like workaround for the Java style IoC. Thanks to that it's possible to inject any Guice manageable dependences.
```
@Singleton
class FileUploadServices @Inject() (config: Configuration) {
```
Then the threshold check is implemented by `hasReachedUploadThreshold`
Supporting evidence:
```
  implicit object SupportingEvidenceUpload extends FileUpload[SupportingEvidencesAnswer] {

    private lazy val max: Int =
      config.underlying.getInt(s"microservice.services.upscan-initiate.max-uploads")

    def hasReachedUploadThreshold(maybeAnswer: Option[SupportingEvidencesAnswer]): Boolean =
      maybeAnswer.exists(_.length >= max)
...
```
Scheduled document:
 ```
   implicit object ScheduledDocumentUpload extends FileUpload[ScheduledDocument] {

    def hasReachedUploadThreshold(maybeAnswer: Option[ScheduledDocument]): Boolean =
      maybeAnswer.isDefined
 ```
 To import all those instances must import fileUploadServices self:
 ```
 class SupportingEvidenceController @Inject() (
  fileUploadServices: FileUploadServices
) {

  import fileUploadServices._
 ```
Hope that make sense